common/src/main/scala/org/neo4j/spark/converter/TypeConverter.scala

```diff
@@ -0,0 +1,135 @@
+  ))
+}
+
+class CypherToSparkTypeConverter extends TypeConverter[String, DataType] {
+  override def convert(sourceType: String, value: Any = null): DataType = sourceType
+    .replaceAll(cleanTerms, "") match {
+    case "Node" | "Relationship" => if (value != null) value.asInstanceOf[Entity].toStruct else DataTypes.NullType
+    case "NodeArray" | "RelationshipArray" => if (value != null) DataTypes.createArrayType(value.asInstanceOf[Entity].toStruct) else DataTypes.NullType
+    case "Boolean" => DataTypes.BooleanType
+    case "Long" => DataTypes.LongType
+    case "Double" => DataTypes.DoubleType
+    case "Point" => pointType
+    case "DateTime" | "ZonedDateTime" | "LocalDateTime" => DataTypes.TimestampType
+    case "Time" => timeType
+    case "Date" => DataTypes.DateType
+    case "Duration" => durationType
+    case "Map" => {
+      val valueType = if (value == null) {
+        DataTypes.NullType
+      } else {
+        val map = value.asInstanceOf[java.util.Map[String, AnyRef]].asScala
+        val types = map.values
+          .map(normalizedClassName)
+          .toSet
+        if (types.size == 1) convert(types.head, map.values.head) else DataTypes.StringType
+      }
+      DataTypes.createMapType(DataTypes.StringType, valueType)
+    }
+    case "Array" => {
+      val valueType = if (value == null) {
+        DataTypes.NullType
+      } else {
+        val list = value.asInstanceOf[java.util.List[AnyRef]].asScala
+        val types = list
+          .map(normalizedClassName)
+          .toSet
+        if (types.size == 1) convert(types.head, list.head) else DataTypes.StringType
+      }
+      DataTypes.createArrayType(valueType)
+    }
+    // These are from APOC
+    case "StringArray" => DataTypes.createArrayType(DataTypes.StringType)
+    case "LongArray" => DataTypes.createArrayType(DataTypes.LongType)
+    case "DoubleArray" => DataTypes.createArrayType(DataTypes.DoubleType)
+    case "BooleanArray" => DataTypes.createArrayType(DataTypes.BooleanType)
+    case "PointArray" => DataTypes.createArrayType(pointType)
+    case "DateTimeArray" => DataTypes.createArrayType(DataTypes.TimestampType)
+    case "TimeArray" => DataTypes.createArrayType(timeType)
+    case "DateArray" => DataTypes.createArrayType(DataTypes.DateType)
+    case "DurationArray" => DataTypes.createArrayType(durationType)
+    // Default is String
+    case _ => DataTypes.StringType
+  }
+}
+
+object SparkToCypherTypeConverter {
+  def apply(): SparkToCypherTypeConverter = new SparkToCypherTypeConverter()
+
+  private val mapping: Map[DataType, String] = Map(
+    DataTypes.BooleanType -> "BOOLEAN",
+    DataTypes.StringType -> "STRING",
+    DataTypes.IntegerType -> "INTEGER",
+    DataTypes.LongType -> "INTEGER",
+    DataTypes.FloatType -> "FLOAT",
+    DataTypes.DoubleType -> "FLOAT",
+    DataTypes.DateType -> "DATE",
+    DataTypes.TimestampType -> "LOCAL DATETIME",
+//    DataTypes.TimestampNTZType -> "ZONED DATETIME",
```

Comment:
maybe remove these?

Reply:
I left as reminder, because it's dependent from how we want to manage dates. If you want I can remove it

Reply:
let's open an issue instead, wdyt?

Reply:
Done